File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-5.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        int l = 0;
        int r = nums.size() - 1;

        while (l < r) {
            int i = (l + r) / 2;

            if (nums[i] < nums[r]) {
                r = i;
            }
            else {
                l = i + 1;
            }
        }
        return nums[l];
    }
};

File: Data Structures & Algorithms/house-robber/submission-5.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        memo = [-1] * n
        
        def dfs(i):
            if i >= n:
                return 0
            if memo[i] == -1:
                memo[i] = max(nums[i] + dfs(i + 2), dfs(i + 1))
            
            return memo[i]
        
        return dfs(0)